MINOR: Fix a few blocking calls in PlaintextConsumerTest

hachikuji opened the pull request:
We've been seeing some hanging builds recently (see KAFKA-7553). Consistently the culprit seems to be a test case in PlaintextConsumerTest. This patch doesn't fix the underlying issue, but it eliminates a few places where these test cases could block:

1. It replaces several calls to the deprecated `poll(long)` which can block indefinitely in the worst case with `poll(Duration)` which respects the timeout.
2. It also fixes a consume utility in `TestUtils` which can block for a long time depending on the number of records that are expected to be consumed.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

hachikuji committed:
MINOR: Fix a few blocking calls in PlaintextConsumerTest

cmccabe commented:
Thanks @hachikuji.  Looks like a good cleanup.

Can you add a comment to the git commit explaining that because we use the new KIP-266 API now, there is no longer a chance that the test might block indefinitely looking for an assignment?

When using `waitUntilTrue` together with `poll`, we should probably just set `pause` to 0, right?  We'd rather wait inside the `poll` operation than inside a `sleep` operation.

LGTM aside from that.

hachikuji commented:
@cmccabe Thanks for reviewing. Good call on `pause`. I have updated the patch. I also updated the description above with more detail as you suggested.

hachikuji commented:
Sorry, actually, I realized that we need to use `pause` everywhere when we poll the consumer with `waitUntil`. I will add a utility and update the patch.

hachikuji committed:
Add utility methods for consumer poll patterns

cmccabe commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -1458,14 +1393,12 @@ class PlaintextConsumerTest extends BaseConsumerTest {
     val fetchLead0 = consumer.metrics.get(new MetricName("records-lead", "consumer-fetch-manager-metrics", "", tags1))
     assertNotNull(fetchLead0)
-    assertTrue(s"The lead should be ${records.count}", fetchLead0.metricValue() == records.count)
+    assertEquals(s"The lead should be ${records.count}", records.count.toDouble, fetchLead0.metricValue())
```
Can we compare these as an integer type rather than floating point?  It is a number of records, right?

ijuma replied:
The metric value is a double though so you don't want to truncate it in case there's a bug.

cmccabe replied:
Makes sense.  Let's leave it as a double, then

cmccabe commented:
Looks good.  The new utility methods look useful.  I left a comment related to comparing floating point values.  LGTM aside from that.

ijuma commented on core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:
```diff
@@ -183,26 +184,22 @@ abstract class BaseConsumerTest extends IntegrationTestHarness {
-      if (iters > maxIters)
-        throw new IllegalStateException("Failed to consume the expected records after " + iters + " iterations.")
-      iters += 1
+    def pollCondition(polledRecords: ConsumerRecords[K, V]): Boolean = {
```
Nit: it's confusing to call this type of function `condition` given that they often involve side-effects. It's an existing issue since that's the parameter name of `waitUntilTrue`. Maybe `action` would be better?

ijuma commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -423,35 +406,26 @@ class PlaintextConsumerTest extends BaseConsumerTest {
     consumer.subscribe(pattern1, new TestConsumerReassignmentListener)
     consumer.poll(50)
```
Is it intentional that this one still uses the deprecated overload?

hachikuji replied:
Not unintentional anyway. The calls that were using a looping pattern could be transformed without any thinking :wink:. Let me take a quick look at the remaining ones and see which ones are easy to replace. This particular call looks totally unnecessary.

ijuma reviewed (commented):
Nice clean-up. I assume we still have some tests that exercise the deprecated method?

hachikuji committed:
Remove superfluous poll() calls

hachikuji committed:
Prefer poll(Duration) and add test case for deprecated behavior

ijuma commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -1304,7 +1296,6 @@ class PlaintextConsumerTest extends BaseConsumerTest {
     consumer.assign(List(tp, tp2).asJava)
 
     // Need to poll to join the group
-    consumer.poll(50)
```
The comment is stale now.

ijuma commented on core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:
```diff
@@ -169,14 +178,14 @@ class PlaintextConsumerTest extends BaseConsumerTest {
     consumer.subscribe(List(topic).asJava, listener)
 
     // poll once to get the initial assignment
```
Seems inaccurate now?

hachikuji committed:
Fix stale comments

hachikuji committed:
Condition -> Action

hachikuji commented:
retest this please

cmccabe commented:
LGTM

hachikuji commented:
I'm going to go ahead and merge. There is still obviously some build instability, but the latest jdk11 failure did trigger the failure we were expecting to see. That validates the changes. See https://issues.apache.org/jira/browse/KAFKA-7553 for an update. The other failures appear to be the usual flaky tests and some unrelated streams failures.
